tests/admin/views/test_rules.py

```diff
@@ -2834,3 +2834,14 @@ def testCannotRevokeOtherUsersSignoffAsNormalUser(self):
         ret = self._delete("/scheduled_changes/rules/3/signoffs", username="julie", qs={"username": "mary"})
         self.assertEqual(ret.status_code, 403, ret.get_data())
         self.assertEqual(ret.mimetype, "application/json")
+
+    def testScheduledChangeWithDuplicateAlias(self):
```

Comment:
This test is not quite right, unfortunately. It is testing the `/rules` view - which is already covered by https://github.com/mozilla-releng/balrog/blob/6493a9e2eae9e959fc69e7e2fd93b6e91ed16ba3/tests/admin/views/test_rules.py#L540.

You should be adding at least two tests:
- One that validates a POST to `/scheduled_changes/rules`, and one that validates a POST to `/scheduled_changes/rules/:sc_id`. The first could be modeled after https://github.com/mozilla-releng/balrog/blob/6493a9e2eae9e959fc69e7e2fd93b6e91ed16ba3/tests/admin/views/test_rules.py#L2161, the second could be modeled after https://github.com/mozilla-releng/balrog/blob/6493a9e2eae9e959fc69e7e2fd93b6e91ed16ba3/tests/admin/views/test_rules.py#L2251

In both cases the request should a result in a 400 if any of the following are true:
- The alias exists in any row in the `rules` table
- The alias exists in any row in the `rule_scheduled_changes` table that meets both the following conditions:
  - `complete` is False
  - `change_type` is `insert` or `update`

(This may require additional changes to `rules.py` as well - please let me know if you have any questions.)

Reply:
I'll start working on updating the test :+1:

Reply:
@bhearsum I have a doubt regarding this - 
How do I select the required data in the rule_scheduled_change table? for eg-
for the rules table we have the `getAlias` function defined to get the rule with the given alias but for rule_scheduled_change I am not able to figure out how should I select the data with a given alias.

Reply:
> @bhearsum I have a doubt regarding this - How do I select the required data in the `rule_scheduled_change` table? for eg- for the `rules` table we have the `getAlias` function defined to get the rule with the given alias but for `rule_scheduled_changes` I am not able to figure out how should I select the data with a given alias.

The `scheduled_changes` table is accessible through `dbo.rules.scheduled_changes` (You can see an example query in the tests https://github.com/mozilla-releng/balrog/blob/6d4620760ca06dc13d12c57ae7eccf6d3233705f/tests/admin/views/test_rules.py#L2012 - and there's many more example of selects in other tests and parts of the codebase).

> and also does the column `scheduled_by` denote the alias for the `rule_scheduled_changes` table?

No -- `scheduled_by` is the person who scheduled the change :). The alias is in the `base_alias` column in the scheduled changes table (apologies -- I forgot about that detail when I posted the earlier comment).

Reply:
Ok got it! :smile: